src/addons/metaservice/src/main/java/ee/ria/xroad/proxy/serverproxy/RestMetadataServiceHandlerImpl.java

```diff
@@ -140,6 +151,89 @@ public void startHandling(HttpServletRequest servletRequest, ProxyMessage reques
+        restResponse.getHeaders().add(new BasicHeader(MimeUtils.HEADER_CONTENT_TYPE, MimeTypes.JSON));
+        MethodListType methodList = OBJECT_FACTORY.createMethodListType();
+        methodList.getService().addAll(ServerConf.getServicesByDescriptionType(
+                requestProxyMessage.getRest().getServiceId().getClientId(), DescriptionType.OPENAPI3));
+        MAPPER.writeValue(restResponseBody, methodList);
+    }
+
+    private void handleAllowedMethods(ProxyMessage requestProxyMessage) throws IOException {
+        restResponse.getHeaders().add(new BasicHeader(MimeUtils.HEADER_CONTENT_TYPE, MimeTypes.JSON));
+        MethodListType methodList = OBJECT_FACTORY.createMethodListType();
+        methodList.getService().addAll(ServerConf.getAllowedServicesByDescriptionType(
+                requestProxyMessage.getRest().getServiceId().getClientId(),
+                requestProxyMessage.getRest().getClientId(),
+                DescriptionType.OPENAPI3));
+        MAPPER.writeValue(restResponseBody, methodList);
+    }
+
+    private void handleGetOpenApi(HttpServletRequest servletRequest, ProxyMessage requestProxyMessage)
+            throws URISyntaxException, IOException {
+        // parse query string
+        String fullURL = getFullURL(servletRequest) + "?" + requestProxyMessage.getRest().getQuery();
+        log.trace("fullURL={}", fullURL);
+        List<NameValuePair> pairs = URLEncodedUtils.parse(new URI(fullURL), Charset.forName("UTF-8"));
+        String targetServiceCode = null;
+        for (NameValuePair pair : pairs) {
+            log.trace("{} : {}", pair.getName(), pair.getValue());
+            if (pair.getName().equalsIgnoreCase(QUERY_PARAM_SERVICECODE)) {
+                targetServiceCode = pair.getValue();
+            }
+        }
+
+        if (targetServiceCode == null || targetServiceCode.isEmpty()) {
+            throw new CodedException(X_INVALID_REQUEST,
+                    "Missing serviceCode in message body");
+        }
+
+        ServiceId targetServiceId = ServiceId.create(requestProxyMessage.getRest().getServiceId().getClientId(),
+                targetServiceCode);
+        log.trace("targetServiceId={}", targetServiceId);
+
+        DescriptionType descriptionType = ServerConf.getDescriptionType(targetServiceId);
+        if (descriptionType == null) {
+            throw new CodedException(X_INTERNAL_ERROR,
+                    String.format("Service not found: %s", targetServiceId.toString()));
+        }
+        if (descriptionType != DescriptionType.OPENAPI3_DESCRIPTION) {
+            throw new CodedException(X_INTERNAL_ERROR,
+                    String.format("Invalid service type: %s", descriptionType.toString()));
+        }
+
+        String serviceDescriptionURL = ServerConf.getServiceDescriptionURL(targetServiceId);
+        URL url = new URL(serviceDescriptionURL);
+
+        byte[] buffer = new byte[BUFFER_SIZE_BYTES];
+        URLConnection urlConnection = url.openConnection();
```

Comment:
If this is a https URL, the default URL connection won't trust the internal server certificates, and it won't send the security server internal certificate if requested --> openapi requests behave differently from normal service requests. see ee.ria.xroad.proxy.serverproxy.MetadataServiceHandlerImpl#getWsdl

Reply:
Switched to HttpClient